apps/shell/utils/helpers.js

```diff
@@ -1,10 +1,67 @@
-const minimatch = require('minimatch')
+const minimatch = require('minimatch');
+const glob      = require('glob');
+const path      = require('path');
+const yaml      = require('js-yaml');
+const fs        = require('fs');
 
 function hostInAllowList(allowlist, host) {
   allowlist = Array.from(allowlist);
   return allowlist.some((value) => minimatch(host, value))
 }
 
+/**
+ * If the url has a match with host_path_rx, the respective capture groups are returned
+ * 
+ * @param {string} url request url
+ * @param {string} default_sshhost returned hostname if default
+ * 
+ * @return {Array}
+ */
+function hostAndDirFromURL(url, default_sshhost){
+  const host_path_rx = '/ssh/([^\\/\\?]+)([^\\?]+)?(\\?.*)?$';
+  let match = url.match(host_path_rx), 
+  hostname = null, 
+  directory = null;
+
+  if (match) {
+    hostname = match[1] === "default" ? default_sshhost : match[1];
+    directory = match[2] ? decodeURIComponent(match[2]) : null;
+  }
+  return [hostname, directory];
+}
+
+class HostAllowlist{
+  constructor(ood_sshhost_allowlist, clusters_d_path, ood_default_sshhost){
+    this.allowlist = ood_sshhost_allowlist ? new Set(ood_sshhost_allowlist.split(':')) : new Set();
+
+    //Filter through cluster configs, return array of hashes with host and default
+    let cluster_sshhosts = glob.sync(path.join((clusters_d_path || '/etc/ood/config/clusters.d'), '*.y*ml'))
+    .map(yml => yaml.safeLoad(fs.readFileSync(yml)))
```

Comment:
Is it possible for this to raise an exception:

    .map(yml => yaml.safeLoad(fs.readFileSync(yml)))

I think if so we should at least rescue from that. Might want to also test if there is a cluster config that exists but the user does not have read access to, which is a legitimate case. So maybe we should filter the cluster config files first for those that the user has read access to. before the `.map`

Reply:
Do you think something like a try-catch would work better? 
In the [https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback](documentation), they recommend reading the file and handling any exceptions that occur. I'm not sure the if the synchronous version has a call back though - need to do more research into it.